fl4health/tasks/autoencoder_trainer.py

```diff
@@ -0,0 +1,182 @@
+from pathlib import Path
+from typing import Callable, Dict, Tuple
+
+import torchvision.transforms as transforms
+from torch.nn.modules.loss import _Loss
+from torch.utils.data import DataLoader
+
+from fl4health.processing.VAE_dim_reduction import ClientConditionedProcessor, VAEProcessor
+from fl4health.processing.VAE_training import AETransformer
+from fl4health.tasks.loss import VAE_loss
+from fl4health.utils.sampler import LabelBasedSampler
+
+
+class VAETrainer:
+    """This Trainer class can enhance any client class within this framework by adding the functionality of training
+    a Variational Auto-encoder. Functionalities include prepare_input()
+    which calls the data loader with an specific VAE transformer for VAE training,
+    and reduce_dim() which is used to reduce the dimension of data during the pre-processing step with VAEs.
+    """
+
+    def prepare_input(
+        self, load_data: Callable, data_path: Path, batch_size: int, sampler: LabelBasedSampler, transform: transforms
+    ) -> Tuple[DataLoader, DataLoader, Dict[str, int]]:
+        """Calls the data loading function with the given parameters and adds a transofrmer
+        to convert the data in the suitable format for VAE training i.e. (sample, sample) format.
+
+        Args:
+            load_data (Callable): Function to load the data.
```

Comment:
IMHO Ideally it would be best not to have this load_data function the user has to pass. It kind of constrains and complicates the construction of the dataloaders. It also presumes that underlying dataset is a `BaseDataset` which we have defined which hasn't been the case in all examples we have been doing and the flamby experiments. 

If possible, take as input the train_loader and val_loader and we can add the transformation in. This allows the user to construct the dataloader in anyway they see fit and not have to worry about how we add the transformation in. 

I don't know how feasible this would be, and it would probably take a bit of thinking to accomplish so feel free to put a ticket in clickup to explore this in future.

Reply:
If we put a ticket in clickup, for now it would be good to add details to the documentation about the input and output you expect for the callable load data, and other information like the fact that its expected is the underlying Dataset is derived from BaseDataset (because we assume the name of certain attributes).

Reply:
Thank you John for this comment! This was actually one of the main challenges that I had. I am happy you pointed that out.
It seems there is no way we can change the data transformers (or add to them) after the data_loaders are created. Or if there is a way, I couldn't find it. I also believe that passing the load_data function would limit the code in many ways. I did that since I wanted to do all the (C)VAE-related codings at the library level, so that the user doesn't need to know which data transformers should be created or how they should be passed to the load_data function.

After your suggestion, I also thought about it again and decided to go with the less constraining and complicated design. But in this new design, users have to set the data transformer for the load_data function themselves. I think this makes sense and is not too bad hopefully.
 
As a better solution, we probably would need to redesign the load_data such that the data transformers can be changed in various modules. This probably needs its own ticket and we can explore it later.